### ISSUE TITLE
[bug] GE dimmers/switches do not reflect double taps and take the device dim out of sync in HA

### ISSUE THREAD

**Issue by blhoward2** [zwave-js/zwavejs2mqtt#116]
**Version**

Build/Run method
- [X] Docker
- [ ] PKG
- [ ] Manually built (git clone - npm install - npm run build ) 

zwavejs2mqtt version: 1.2.3

**Describe the bug**

GE dimmer/switches are capable of double tap events. In zwave2mqtt these show as events under the node, with 255 indicating double tap up and 0 indicating double tap down. (My ozw device files were set to ignore the command mapping for class 32 from qt-openzwave and I carried this over when I switched so I don't know if it was necessary. 

These are the steps to make it work under qt-openzwave: https://community.home-assistant.io/t/howto-add-double-tap-support-for-ge-switches-and-dimmers-in-ozw-beta/228871/16. 

The original qt-openzwave issue which has more detailed info is at: https://github.com/OpenZWave/qt-openzwave/issues/60.

Double tap up in zwavejs2mqtt now shows:

```
zwavejs2mqtt    | 2020-12-28 23:32:24.898 INFO ZWAVE: Node 3: value updated: 38-0-currentValue 50 => 255

```
(The current value goes to 255 and HA thinks the device is now at 100%, but the physical dim has not changed. **They are now out of sync**.)

Both regular down and double-tap down show 0 so I don't see a way to differentiate those:

`zwavejs2mqtt    | 2020-12-28 23:21:25.052 INFO ZWAVE: Node 3: value updated: 38-0-currentValue 50 => 0`

**To Reproduce**
1. Double click up, or double click down

**Expected behavior**
Double tap up should generate an mqtt event or message that can be acted upon, if not a sensor within HA. In zwave2mqtt I just subscribed to the topic in Appdaemon.

**Additional context**

**Comment by blhoward2** [comment]
Note that these are older zwave plus models of the GE14294, not the newer models with Central Scene states.

**Comment by blhoward2** [comment]
More info from logs.

Double tap up:

```

04:43:19.204 DRIVER « [Node 003] [REQ] [ApplicationCommand]
                      └─[BasicCCSet]
                          target value: 255
04:43:19.209 CNTRLR   [Node 003] treating BasicCC Set as a report
04:43:19.210 CNTRLR   [Node 003] [~] [Multilevel Switch] currentValue: 50 => 255        [Endpoint 0]
04:43:19.285 SERIAL « 0x01080004006702820115                                              (10 bytes)
04:43:19.286 DRIVER   Dropping message because it could not be deserialized: The command class Hail 
                      is not implemented
04:43:19.287 SERIAL » [ACK]                                                                   (0x06)
```


Double tap down:

```
04:43:59.988 DRIVER « [Node 003] [REQ] [ApplicationCommand]
                      └─[BasicCCSet]
                          target value: 0
04:43:59.991 CNTRLR   [Node 003] treating BasicCC Set as a report
04:43:59.992 CNTRLR   [Node 003] [~] [Multilevel Switch] currentValue: 255 => 0         [Endpoint 0]
04:44:03.026 SERIAL « 0x01080004006702820115                                              (10 bytes)
04:44:03.027 DRIVER   Dropping message because it could not be deserialized: The command class Hail 
                      is not implemented
04:44:03.028 SERIAL » [ACK]                  
```

**Comment by robertsLando** [comment]
@blhoward2 Could you also show me the console logs of zjs2m?

**Comment by blhoward2** [comment]
I can confirm that the same issue persists on the most recent dev image, pulled today. 

Double-tap up (from a prior double-tap up state, hence 255-->255):

![Screen Shot 2021-01-02 at 10 13 38 PM](https://user-images.githubusercontent.com/33612110/103471023-a8ce2b80-4d48-11eb-9303-8abb1f56e357.png)

Double-tap up from off, just in case being in a former double-tap state affects it (note this one is out of order on the console log):
![Screen Shot 2021-01-02 at 10 16 16 PM](https://user-images.githubusercontent.com/33612110/103471044-ff3b6a00-4d48-11eb-92c4-31476e4561ca.png)


Double-tap down:
![Screen Shot 2021-01-02 at 10 14 02 PM](https://user-images.githubusercontent.com/33612110/103471030-bdaabf00-4d48-11eb-8746-6a44cec21ce0.png)

For comparison, off to normal on:
![Screen Shot 2021-01-02 at 10 15 11 PM](https://user-images.githubusercontent.com/33612110/103471033-cef3cb80-4d48-11eb-9762-7c910317e0dd.png)

For comparison, on to normal off:
![Screen Shot 2021-01-02 at 10 15 43 PM](https://user-images.githubusercontent.com/33612110/103471039-e29f3200-4d48-11eb-801c-e222d684e2a0.png)

Console log:

![Screen Shot 2021-01-02 at 10 17 08 PM](https://user-images.githubusercontent.com/33612110/103471049-0f534980-4d49-11eb-90da-dcc00e2fecc6.png)

**Comment by blhoward2** [comment]
I opened an upstream bug as I believe this results from this: https://github.com/zwave-js/node-zwave-js/issues/388

**Comment by towerhand** [comment]
I think this works in out of the box in zwave2mqtt without the user having to make any special changes unlike in qtozw.

**Comment by blhoward2** [comment]
I’m working with AlCalzone in the upstream bug report. He’s already committed a change to try to provide a compat flag to fix this. The devices are non-compliant with the standard and when the standard itself became publicly available the projects made changes to how things worked to comport with the standard. Hence why it works in the older projects and not the newer ones. 

I’ve left this bug report open as I’m unsure of downstream changes will also be required.

**Comment by towerhand** [comment]
Thanks, I’m keeping my eye on this. I’ll switch over from z2m once this issue is resolved.

**Comment by blhoward2** [comment]
For now I’ve just removed the controller from group 3 and am relying solely on direct association and those devices calling in their state change. It mostly works fine that way except for a few odd devices I use to refresh after a double-tap event for reliability.

**Comment by robertsLando** [comment]
@AlCalzone

**Comment by AlCalzone** [comment]
@robertsLando We're already on it and need your help:
https://github.com/zwave-js/node-zwave-js/issues/1277

**Comment by robertsLando** [comment]
@AlCalzone Sorry, catching up missed notifications :smile:

**Comment by blhoward2** [comment]
Ok, the changes made upstream now work (I'll do a PR to add the compat flag to the device files.) I'm receiving the basic set events on command class 32 for double-up and double-down. I tried refreshing the node info but no sensors are added within the ui. What do we need to change to define a sensor or something so HA more readily receives these events?

**Comment by towerhand** [comment]
@blhoward2 Thanks for the heads up. I'll be installing zj2m in the next release if this is working properly. Hopefully the basic cc doesn't get activated during HA restart like Kpine mentioned in the other issue. Glad to see bug getting addressed so quickly, main reason I kept using z2m after a short trial with qtozw.

**Comment by blhoward2** [comment]
PR for the final change upstream submitted (modifying the device files to take advantage of the new compat flag). https://github.com/zwave-js/node-zwave-js/pull/1303

**Comment by towerhand** [comment]
Would this be available if I install zj2m wiith docker using the dev tag?

**Comment by blhoward2** [comment]
Not yet, no. @AlCalzone can correct me if I’m wrong but I believe the patch is built on top of zwave 6.0, which was incorporated by the pending #84 PR.

**Comment by towerhand** [comment]
Thanks @blhoward2, I'll wait until that issue is resolved before jumping in the testing wagon.

**Comment by AlCalzone** [comment]
@all you could try if the double config issue is gone with https://github.com/zwave-js/zwavejs2mqtt/pull/181. I must admit I have no clue how to run it with Docker for Windows.

**Comment by blhoward2** [comment]
I’ll do so tonight.

**Comment by robertsLando** [comment]
I think this has been opened in node-zwave-js. Closing this here

**Comment by blhoward2** [comment]
Sort of. The node fix causes the values to be sent. We still do nothing with setting up devices in HA for those, which is why I opened this one.

**Comment by towerhand** [comment]
> > > taking back what i just wrote, I do see this in the handle value notification function, which appears to be making an event... is that in MQTT?
> > 
> > 
> > If you enable the flag `publiush events` in gateway all zwavejs events are published to mqtt: https://zwave-js.github.io/zwavejs2mqtt/#/guide/mqtt?id=zwave-events
> 
> Yeah, I did that intentionally, but HA doesn’t see those and it would be difficult to define a new sensor under the device watching all events.
> 
> We need a value under the node that works like the scene one now does where it resets to 0. We tried to figure out last night why we aren’t getting a 32 command class under the node at all now (we did when sending value events), and I posted in the PR itself.
Moving to this issue since is probably more relevant here.

I installed zjs2m this morning to test and see how things are evolving and try to provide info from a regular user point of view.

I can see the events getting triggered with the console and with mqtt explorer in the event topic in the root topic per the docs (<mqtt_prefix>/_EVENTS_/ZWAVE_GATEWAY-). In the old z2m it would publish them in the node topic as well and made everything a lot easier for regular user like me (<mqtt_prefix>/node_id/event) to set up automations to watch the event for an specific device.

**Comment by blhoward2** [comment]
If you go to this discussion on the node-Zwave-js project, this is evolving. As of the image you should get a command class 32 after you double tap that shows the value. We’re working to change this to a value notification as presently HA won’t see subsequent double-taps of the same value

**Comment by blhoward2** [comment]
@robertsLando We now have this sorted at the node level, and it works perfectly when I define a manual sensor. How do I go about defining a sensor so it is automatically created? Can it be done when it first gets a double_tap event (when the mqtt topic is created)? Or would a user now need to custom add this because there isn't a way to prevent it being doing on all switches?

```
{
  "type": "sensor",
  "object_id": "double_tap",
  "discovery_payload": {
    "state_topic": "homeassistant/3/32/0/value",
    "value_template": "{{ value_json.value}}",
    "json_attributes_topic": "homeassistant/3/32/0/value",
    "device": {
      "identifiers": [
        "zwavejs2mqtt_0xeb87ad4a_node3"
      ],
      "manufacturer": "Jasco Products",
      "model": "In-Wall Dimmer Switch (GE 14294 (ZW3005))",
      "name": "nodeID_3",
      "sw_version": "5.29"
    },
    "name": "nodeID_3_double_tap",
    "unique_id": "zwavejs2mqtt_0xeb87ad4a_3-32-0-double_tap"
  },
  "discoveryTopic": "sensor/nodeID_3/double_tap/config",
  "values": [
    "32-0-value"
  ],
  "persistent": false,
  "ignoreDiscovery": false,
  "id": "sensor_double_tap"
}
```

**Comment by AlCalzone** [comment]
heads up: I'm going to rename the /value to /event to avoid confusion with target and currentValue.

**Comment by blhoward2** [comment]
> heads up: I'm going to rename the /value to /event to avoid confusion with target and currentValue.

Makes sense to me

**Comment by blhoward2** [comment]
> heads up: I **have renamed** the /value to /event to avoid confusion with target and currentValue.

Do my changes above capture the change, or is it now 32-0-event too?

**Comment by AlCalzone** [comment]
It is now `32-0-event`

**Comment by robertsLando** [comment]
Can we close this so?

**Comment by blhoward2** [comment]
> Can we close this so?

I posted two questions for you a few messages up. Can we auto-create a sensor for this or would a user have to define their own? And there is still a bug with the log that causes it to say undefined.

**Comment by robertsLando** [comment]
Sorry missed them, too much notifications :cry: 

> We also get an undefined message in the logging:

undefined is the previous value, for a notification it is undefined, if you want I could remove the undefined from the log but it's not wrong as it says: I was undefined and now I'm 255

> Can we auto-create a sensor for this or would a user have to define their own?

Absolutely, to cleanup some stuff would you open a dedicated issue? Like `[feat] create HASS sensor for CC 32 event`

**Comment by AlCalzone** [comment]
In the context of value notifications (stateless), displaying a transition from a previous value makes no sense IMO.

**Comment by robertsLando** [comment]
Ok will remove it so, no problem

**Comment by blhoward2** [comment]
> Absolutely, to cleanup some stuff would you open a dedicated issue? Like `[feat] create HASS sensor for CC 32 event`

Done

**Comment by robertsLando** [comment]
The log is fixed on master closing this in favor of #249
